### PR TITLE
Add uniform and tests for opengraph, microdata and microformat

### PR DESCRIPTION
**Pull request by Kebniss**
Still work in progress, need to define wether to add rdfa normalization here or to another pull request

### CODE REVIEW

**Comment by lopuhin** [comment]
Re data loss when ``uniform=True`` is passed - is it possible to avoid it? Any examples of it?

File: extruct/uniform.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    return res
+
+
+def flatten_dict(d, schema_context, add_context):
+    typ = d.get('type', None)
+    if not typ:
+        return d
+
+    if isinstance(typ, list):
+        assert len(typ), "Multiple types not supported, storing the first one"
```
[the remark this answers is not in the document]
Comment (Kebniss) [reply extruct/uniform.py]: Fine by me. This still comes from Rolando's code so I don't have a very clear idea of when it's useful, I guess it's for microformat. TBH if type is a list with multiple elements we could simply copy it. Maybe a single element was needed to create 'merged' in dse-microdata

File: extruct/uniform.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
+from six.moves.urllib.parse import urlparse, urljoin
+
+def _uopengraph(extracted):
+    out = []
+    for obj in extracted:
+        # FIXME: Handle objects in order.
+        # FIXME: Handle arrays (repeated properties).
+        flattened = dict(obj['properties'])
```
[the remark this answers is not in the document]
Comment (Kebniss) [reply extruct/uniform.py]: yes OpenGraphExtractor returns a dict with two keys: namespace and properties, see [here](https://github.com/scrapinghub/extruct/blob/master/extruct/opengraph.py#L26)

**Comment by Kebniss** [comment]
I think I misunderstood the FIXME notes in _uopengraph. It does not seem there is a problem with missing content

File: extruct/tool.py [hunk 3]
```diff
@@ -25,6 +29,15 @@ def main(args=None):
...
+        help='''If True uniform output format of all syntaxes to a list of dicts.
+                Returned dicts structure:
+                {'@context': 'http://example.com', 
+                 '@type': 'example_type',
+                 /* All other the properties in keys here */
+                 }''')
+    arg('--schema_context', default='http://schema.org', 
+        help="schema's context for current page")
     args = parser.parse_args(args)
     metadata = metadata_from_url(args.url, args.syntaxes)
```
Comment (lopuhin) [inline extruct/tool.py]: It seems that arguments added above (uniform and schema_context) are not passed into ``metadata_from_url``

File: extruct/uniform.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        out['@type'] = typ
+    
+    if add_context:
+        out['@context'] = context
+
+    props = out.pop('properties', {})
+    for field, value in props.items():
+        if isinstance(value, dict):
+            value = flatten_dict(value, schema_context, False)
+        elif isinstance(value,
```
Comment (lopuhin) [inline extruct/uniform.py]: this newline is not needed?

File: extruct/uniform.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Kebniss) [reply extruct/uniform.py]: Yeah it seems so. Probably are just outdated comments. I think it's safe to remove them

File: extruct/uniform.py (hunk 4 above)
Comment (Kebniss) [reply extruct/uniform.py]: ups

**Comment by kmike** [comment]
Thanks @Kebniss!

**Pull request merged**